subsys/bluetooth/scan.c

```diff
@@ -283,6 +319,141 @@ static void notify_connecting_error(struct bt_scan_device_info *device_info)
 	}
 }
 
+static bool blacklist_device_check(const bt_addr_le_t *addr)
+{
+	bool blacklist_device = false;
+
+	k_mutex_lock(&blacklist_mutex, K_FOREVER);
+
+	for (size_t i = 0; i < bt_scan.blacklist.cnt; i++) {
+		if (bt_addr_le_cmp(&bt_scan.blacklist.addr[i], addr) == 0) {
+			blacklist_device = true;
+		}
+	}
+
+	k_mutex_unlock(&blacklist_mutex);
+
+	return blacklist_device;
+}
+
+static void attempts_filter_force_add(struct conn_attempts_filter *filter,
+				      const bt_addr_le_t *addr)
+{
+	/* Overwrite the oldest device */
+	memset(filter->oldest_device, 0, sizeof(*filter->oldest_device));
+	bt_addr_le_copy(&filter->oldest_device->addr, addr);
+
+	if (filter->oldest_device == &filter->device[filter->count - 1]) {
+		filter->oldest_device = filter->device;
+
+		return;
+	}
+
+	filter->oldest_device++;
+}
+
+static void scan_attempts_filter_device_add(const bt_addr_le_t *addr)
+{
+	struct conn_attempts_filter *filter = &bt_scan.attempts_filter;
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+
+	k_mutex_lock(&conn_attempts_mutex, K_FOREVER);
+
+	/* Check if device is already in the filter array. */
+	for (size_t i = 0; i < filter->count; i++) {
+		struct conn_attempts_device *device = &filter->device[i];
+
+		if (bt_addr_le_cmp(addr, &device->addr) == 0) {
+			LOG_DBG("Device %s is already in the filter array",
+				log_strdup(addr_str));
+			goto out;
+		}
+	}
+
+	if (filter->count >= ARRAY_SIZE(filter->device)) {
+		LOG_DBG("Force adding %s device filter",
+			log_strdup(addr_str));
+		attempts_filter_force_add(filter, addr);
+	} else {
+		bt_addr_le_copy(&filter->device[filter->count].addr, addr);
+		filter->count++;
+	}
+
+out:
+	k_mutex_unlock(&conn_attempts_mutex);
+}
+
+static void device_conn_attempts_count(struct bt_conn *conn)
+{
+	const bt_addr_le_t *addr = bt_conn_get_dst(conn);
+	struct conn_attempts_filter *filter = &bt_scan.attempts_filter;
+
+	k_mutex_lock(&conn_attempts_mutex, K_FOREVER);
+
+	for (size_t i = 0; i < filter->count; i++) {
+		struct conn_attempts_device *device = &filter->device[i];
+
+		if (bt_addr_le_cmp(addr, &device->addr) == 0) {
+			if (device->attempts < CONFIG_BT_SCAN_CONNECTION_ATTEMPTS_CNT) {
+				device->attempts++;
+
+				break;
+			}
+		}
+	}
+
+	k_mutex_unlock(&conn_attempts_mutex);
+}
+
+static bool conn_attempts_filter_check(const bt_addr_le_t *addr)
```

Comment:
Maybe we could rename it to `conn_attempts_exceeded` and remove the negations?

Reply:
Yes sure. Done